### PR TITLE
Explore rf params

### PR DESCRIPTION
**Pull request by klaricch**
Edits to ancestry inference to include v2 known pops, v3 spike-ins, and v4 spike-ins

### CODE REVIEW

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 1]
```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
...
+
+        # Filter to only pre-determined list of v3 cohorts for the v3 spike-ins
+        joint_qc_meta = joint_qc_meta.filter(
+            v3_spike_projects.contains(joint_qc_meta.v3_meta.v3_project_pop)
+        )
+        joint_qc_meta = joint_qc_meta.filter(
+            v3_spike_projects[joint_qc_meta.v3_meta.v3_project_pop].contains(
+                joint_qc_meta.v3.project_meta.research_project
+            )
+        )
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:267]:
```suggestion

        # Filter to only pre-determined list of v3 cohorts for the v3 spike-ins
        joint_qc_meta = joint_qc_meta.filter(
            v3_spike_projects.contains(joint_qc_meta.v3_meta.v3_project_pop)
        )
        joint_qc_meta = joint_qc_meta.filter(
            v3_spike_projects[joint_qc_meta.v3_meta.v3_project_pop].contains(
                joint_qc_meta.v3_meta.v3_research_project
            )
        )
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 2]
```diff
@@ -65,10 +117,12 @@ def calculate_mislabeled_training(pop_ht: hl.Table, pop_field: str) -> [int, flo
     :param pop_field: Name of field in the Table containing the assigned pop/subpop.
     :return: The number and proportion of mislabeled training samples.
     """
+    logger.info("Calculating mislabeled training samples")
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:120]:
```suggestion
    logger.info("Calculating mislabeled training samples...")
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 3]
```diff
@@ -65,10 +117,12 @@ def calculate_mislabeled_training(pop_ht: hl.Table, pop_field: str) -> [int, flo
...
     n_mislabeled_samples = pop_ht.aggregate(
         hl.agg.count_where(pop_ht.training_pop != pop_ht[pop_field])
     )
 
+    logger.info("Counting where training pop is defined")
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:125]:
```suggestion
    logger.info("Counting where training pop is defined...")
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 4]
```diff
@@ -84,6 +138,8 @@ def prep_ht_for_rf(
     seed: int = 24,
     test: bool = False,
     only_train_on_hgdp_tgp: bool = False,
+    v4_population_spike: List[str] = None,
+    v3_population_spike: List[str] = None,
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:142]:
```suggestion
    v4_population_spike: Optional[List[str]] = None,
    v3_population_spike: Optional[List[str]] = None,
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 5]
```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
     :param seed: Random seed, defaults to 24.
     :param test: Whether RF should run on the test QC MT.
     :param only_train_on_hgdp_tgp: Whether to train RF classifier using only the HGDP and 1KG populations. Default is False.
+    :param v4_population_spike: List of populations to spike into training. Must be in v4_population_spike dictionary. Default is None.
+    :param v3_population_spike: List of populations to spike into training. Must be in v3_population_spike dictionary. Default is None.
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:153]:
```suggestion
    :param v4_population_spike: Optional List of populations to spike into training. Must be in v4_population_spike dictionary. Default is None.
    :param v3_population_spike: Optional List of populations to spike into training. Must be in v3_population_spike dictionary. Default is None.
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 6]
```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
...
     :return Table with input for the random forest.
     """
     pop_pca_scores_ht = ancestry_pca_scores(include_unreleasable_samples, test).ht()
+    global joint_qc_meta
+    global v3_meta
+
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:159]: Why is this needed?

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 7]
```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
...
     joint_meta = joint_qc_meta.ht()[pop_pca_scores_ht.key]
 
-    # TODO: Add code for subpopulations
+    hgdp_tgp_outliers = hgdp_tgp_pop_outliers.ht()
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:162]: Add comment about why this is needed and what it is

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 8]
```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
...
+    hgdp_tgp_outliers = hgdp_tgp_outliers.s.collect()
 
-    # Either train the RF with only HGDP and TGP or all known populations and
-    # previously inferred pops in v3
+    # TODO: Add code for subpopulations
+    # Either train the RF with only HGDP and TGP, or HGDP and TGP and all v2 known labels
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:166]: Change this comment to include v3 and v4 options

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 9]
```diff
@@ -84,6 +138,8 @@ def prep_ht_for_rf(
     seed: int = 24,
     test: bool = False,
     only_train_on_hgdp_tgp: bool = False,
+    v4_population_spike: List[str] = None,
+    v3_population_spike: List[str] = None,
 ) -> hl.Table:
     """
     Prepare the PCA scores hail Table for the random forest population assignment runs.
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:145]: Might be good to add a bit more detail to this about the options, like instead of the comment on line 166, put that info here and expand it to include v3 and v4 options

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 10]
```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
...
     """
     pop_pca_scores_ht = ancestry_pca_scores(include_unreleasable_samples, test).ht()
+    global joint_qc_meta
+    global v3_meta
+
     joint_meta = joint_qc_meta.ht()[pop_pca_scores_ht.key]
 
-    # TODO: Add code for subpopulations
+    hgdp_tgp_outliers = hgdp_tgp_pop_outliers.ht()
+    hgdp_tgp_outliers = hgdp_tgp_outliers.s.collect()
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:163]:
```suggestion
    hgdp_tgp_outliers = hl.literal(hgdp_tgp_pop_outliers.ht().s.collect())
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 11]
```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
...
-    # Either train the RF with only HGDP and TGP or all known populations and
-    # previously inferred pops in v3
+    # TODO: Add code for subpopulations
+    # Either train the RF with only HGDP and TGP, or HGDP and TGP and all v2 known labels
     if only_train_on_hgdp_tgp:
         logger.info("Training using HGDP and 1KG samples only...")
         training_pop = hl.or_missing(
-            (joint_meta.v3_meta.v3_subsets.hgdp | joint_meta.v3_meta.v3_subsets.tgp),
+            (joint_meta.v3_meta.v3_subsets.hgdp | joint_meta.v3_meta.v3_subsets.tgp)
+            & (joint_meta.v3_meta.v3_project_pop != "oth")
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:171]: Add comment about why this `oth` exclusion is needed

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 12]
```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
...
+            & (~hl.literal(hgdp_tgp_outliers).contains(pop_pca_scores_ht.s)),
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:172]:
```suggestion
            & ~hgdp_tgp_outliers.contains(pop_pca_scores_ht.s),
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 13]
```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
...
-            .or_missing()  # TODO: Do we want a case for v2 known pops after v3 inferred?
-        )  # TODO: Analysis of where v2_pop does not agree with v3 assigned pop
+        logger.info(
+            "Training using HGDP and 1KG samples and all v2 samples with known pops..."
+        )
+        training_pop = hl.if_else(
+            (joint_meta.v3_meta.v3_subsets.hgdp | joint_meta.v3_meta.v3_subsets.tgp)
+            & (joint_meta.v3_meta.v3_project_pop != "oth")
+            & (~hl.literal(hgdp_tgp_outliers).contains(pop_pca_scores_ht.s)),
+            joint_meta.v3_meta.v3_project_pop,
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:183]: This is all the same as above so you can just move the above outside of the if statement, then make this like an `if include_v2_known_in_training:` instead of having a `only_train_on_hgdp_tgp` parameter. Then do what you do below with 
```            .when(
                hl.is_defined(pop_pca_scores_ht.training_pop),
                pop_pca_scores_ht.training_pop,
            )
```

This will need to be moved up too, or down to the bottom and only change the `training_pop` expression not the `pop_pca_scores_ht` . 
```
    pop_pca_scores_ht = pop_pca_scores_ht.annotate(
        training_pop=training_pop,
        hgdp_or_tgp=joint_meta.v3_meta.v3_subsets.hgdp
        | joint_meta.v3_meta.v3_subsets.tgp,
    )
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 14]
```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
...
+        )
 
     pop_pca_scores_ht = pop_pca_scores_ht.annotate(
         training_pop=training_pop,
         hgdp_or_tgp=joint_meta.v3_meta.v3_subsets.hgdp
         | joint_meta.v3_meta.v3_subsets.tgp,
     )
+    if v4_population_spike:
+        logger.info(
+            "Spiking v4 pops, %s, into the RF training data", v4_population_spike
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:201]:
```suggestion
            "Spiking v4 pops, %s, into the RF training data...", v4_population_spike
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 15]
```diff
@@ -137,6 +292,7 @@ def prep_ht_for_rf(
     # Use the withhold proportion to create PR curves for when the RF removes
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:292]: Im honestly think we should just remove the `withhold_prop` and `max_number_mislabeled_training_samples` and `max_proportion_mislabeled_training_samples` from the code. It's confusing, and we determined it's probably not the right way to do this.

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 16]
```diff
@@ -164,6 +320,8 @@ def assign_pops(
     test: bool = False,
     overwrite: bool = False,
     only_train_on_hgdp_tgp: bool = False,
+    v4_population_spike: List[str] = None,
+    v3_population_spike: List[str] = None,
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:324]:
```suggestion
    v4_population_spike: Optional[List[str]] = None,
    v3_population_spike: Optional[List[str]] = None,
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 17]
```diff
@@ -186,6 +344,8 @@ def assign_pops(
     :param test: Whether running assigment on a test dataset.
     :param overwrite: Whether to overwrite existing files.
     :param only_train_on_hgdp_tgp: Whether to train the RF classifier using only the HGDP and 1KG populations. Defaults to False.
+    :param v4_population_spike: List of v4 populations to spike into the RF. Must be in v4_pop_spike dictionary. Defaults to None.
+    :param v3_population_spike: List of v3 populations to spike into the RF. Must be in v4_pop_spike dictionary. Defaults to None.
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:348]:
```suggestion
    :param v4_population_spike: Optional List of v4 populations to spike into the RF. Must be in v4_pop_spike dictionary. Defaults to None.
    :param v3_population_spike: Optional List of v3 populations to spike into the RF. Must be in v4_pop_spike dictionary. Defaults to None.
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 18]
```diff
@@ -206,19 +366,24 @@ def assign_pops(
     else:
         max_mislabeled = None
 
+    logger.info("Prepping HT for RF.")
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:369]:
```suggestion
    logger.info("Prepping HT for RF...")
```

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 19]
```diff
@@ -243,7 +411,11 @@ def assign_pops(
         else prop_mislabeled_samples
     )
     pop_assignment_iter = 1
-
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:246]: Same thought as above, lets just remove this confusing aspect of the code

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 20]
```diff
@@ -311,12 +495,18 @@ def assign_pops(
...
         min_prob=min_prob,
         include_unreleasable_samples=include_unreleasable_samples,
-        max_mislabeled=max_mislabeled,
         pop_assignment_iterations=pop_assignment_iter,
         pcs=pcs,
-        n_mislabeled_training_samples=n_mislabeled_samples,
-        prop_mislabeled_training_samples=prop_mislabeled_samples,
+        only_train_on_hgdp_tgp=only_train_on_hgdp_tgp,
+        v3_population_spike=v3_population_spike,
+        v4_population_spike=v4_population_spike,
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:502]: I think this won't work if `None`

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 21]
```diff
@@ -367,60 +557,84 @@ def write_pca_results(
...
+                get_pop_ht(
+                    test=test, only_train_on_hgdp_tgp=only_train_on_hgdp_tgp
+                ).path,
+                overwrite=overwrite,
+                _read_if_exists=not overwrite,
+            )
+            pop_ht.transmute(
+                **{f"PC{j}": pop_ht.pca_scores[i] for i, j in enumerate(pop_pcs)}
+            ).export(
+                pop_tsv_path(test=test, only_train_on_hgdp_tgp=only_train_on_hgdp_tgp)
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:626]: Do we actually still need the `pop_tsv_path`?

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 22]
```diff
@@ -505,6 +721,18 @@ def main(args):
         help="Whether to train RF classifier using only the HGDP and TGP populations.",
         action="store_true",
     )
+    parser.add_argument(
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:724]: Does it work to add `choices` here?

**Review by jkgoodrich — CHANGES_REQUESTED** [review]

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 23]
```diff
@@ -367,60 +557,84 @@ def write_pca_results(
...
+                withhold_prop=args.withhold_prop,
+                test=test,
+                overwrite=overwrite,
+                only_train_on_hgdp_tgp=only_train_on_hgdp_tgp,
+                v4_population_spike=args.v4_population_spike,
+                v3_population_spike=args.v3_population_spike,
+            )
+            logger.info("Writing pop ht...")
+            pop_ht = pop_ht.checkpoint(
+                get_pop_ht(
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:617]: I think we need to modify `get_pop_ht` to return a finalized pop HT by default, so even if we had decided to train with only hgdp_tgp, that would only be included in the globals and by default when calling `get_pop_ht` you could still get the hgdp_tgp results if that is what we determined was the final pop training. It helps downstream scripts and future users of the resources if it's obvious what the final choice was

File: gnomad_qc/v4/sample_qc/assign_ancestry.py (hunk 21 above)
Comment (klaricch) [reply gnomad_qc/v4/sample_qc/assign_ancestry.py:626]: I'm not sure what it has been used for in the past, I can get rid of it for now

File: gnomad_qc/v4/sample_qc/assign_ancestry.py (hunk 6 above)
Comment (klaricch) [reply gnomad_qc/v4/sample_qc/assign_ancestry.py:134]: otherwise get UnboundLocalError

File: gnomad_qc/v4/sample_qc/assign_ancestry.py [hunk 24]
```diff
@@ -229,98 +258,25 @@ def assign_pops(
...
-        pop_ht, pop_field
-    )
-
-    mislabeled = (
-        n_mislabeled_samples
-        if max_number_mislabeled_training_samples
-        # TODO: Run analysis on what makes sense as input here, utilize withhold
-        # arg for PR curves
-        else prop_mislabeled_samples
+    pop_ht = pop_ht.checkpoint(
```
Comment (jkgoodrich) [inline gnomad_qc/v4/sample_qc/assign_ancestry.py:261]: Is this needed?

**Review by jkgoodrich — COMMENTED** [review]
One last comment

File: gnomad_qc/v4/sample_qc/assign_ancestry.py (hunk 24 above)
Comment (klaricch) [reply gnomad_qc/v4/sample_qc/assign_ancestry.py:261]: hm i think we can drop it

**Review by jkgoodrich — APPROVED** [review]
LGTM
